src/Build/BackEnd/Components/Logging/LoggingService.cs

```diff
@@ -481,14 +482,40 @@ public ISet<string> WarningsAsMessages
         public bool HasBuildSubmissionLoggedErrors(int submissionId)
         {
             // Warnings as errors are not tracked if the user did not specify to do so
-            if (WarningsAsErrors == null)
+            if (WarningsAsErrors == null && _filterEventSource.WarningsAsErrorsByProject == null && _eventSinkDictionary.Values.All(i => i.WarningsAsErrorsByProject == null))
             {
                 return false;
             }
 
             // Determine if any of the event sinks have logged an error with this submission ID
-            return (_filterEventSource != null && _filterEventSource.BuildSubmissionIdsThatHaveLoggedErrors.Contains(submissionId))
-                || (_eventSinkDictionary != null && _eventSinkDictionary.Values.Any(i => i.BuildSubmissionIdsThatHaveLoggedErrors.Contains(submissionId)));
+            return _filterEventSource != null && _filterEventSource.BuildSubmissionIdsThatHaveLoggedErrors.Contains(submissionId)
+                || _eventSinkDictionary != null && _eventSinkDictionary.Values.Any(i => i.BuildSubmissionIdsThatHaveLoggedErrors.Contains(submissionId));
+        }
+
+        public void AddWarningsAsErrors(int projectInstanceId, ISet<string> codes)
+        {
+            lock (_lockObject)
+            {
+                if (_filterEventSource.WarningsAsErrorsByProject == null)
+                {
+                    _filterEventSource.WarningsAsErrorsByProject = new ConcurrentDictionary<int, ISet<string>>();
+                }
+
+                _filterEventSource.WarningsAsErrorsByProject.Add(projectInstanceId, new HashSet<string>(codes, StringComparer.OrdinalIgnoreCase));
+            }
+        }
+
+        public void AddWarningsAsMessages(int projectInstanceId, ISet<string> codes)
+        {
+            lock (_lockObject)
```

Comment:
I tried this out but hit a snag.  I'm using the state of `null` as an indicator if any warnings should be treated as an error/message.  If I use a `Lazy<T>`, then I would have to expose it as such so that the caller can determine if `HasValue` is `true`.  Leaving it the way it is, the property remains of type `IDictionary` and a simple null check is done.  If its okay to have `IBuildEventSink` have a property that is of type `Lazy<T>`, then I'll do it.